src/global.cpp

```diff
@@ -36,27 +36,108 @@ int symPACK_Rank(int * rank){
+}
+ 
+extern "C"
+void symPACK_cuda_setup(symPACK::symPACKOptions optionsFact) {
+  if (optionsFact.gpu_alloc_size != 0) {
+    symPACK::gpu_alloc_size = optionsFact.gpu_alloc_size;  
+  }
+
+  symPACK::fallback_type = optionsFact.fallback_type; 
+
+  symPACK::gpu_allocator = upcxx::make_gpu_allocator<upcxx::gpu_default_device>(symPACK::gpu_alloc_size);
+
+  int gpu_id = symPACK::gpu_allocator.device_id();
+  cudaSetDevice(gpu_id);
+  
+  cublasHandle_t handle;
+  symPACK::cublas_handler= handle; 
+  CUBLAS_ERROR_CHECK(cublasCreate(&symPACK::cublas_handler));
+
+  cusolverDnHandle_t cusolver_handle;
+  symPACK::cusolver_handler = cusolver_handle;
+  CUSOLVER_ERROR_CHECK(cusolverDnCreate(&symPACK::cusolver_handler));
+  
+  symPACK::gpu_block_limit = optionsFact.gpu_block_limit;
+  symPACK::trsm_limit = optionsFact.trsm_limit;
+  symPACK::potrf_limit = optionsFact.potrf_limit;
+  symPACK::gemm_limit = optionsFact.gemm_limit;
+  symPACK::syrk_limit = optionsFact.syrk_limit;
+  symPACK::gpu_solve = optionsFact.gpu_solve;
+  symPACK::gpu_verbose = optionsFact.gpu_verbose;
+  
+  if (optionsFact.gpu_verbose) {
+    if (upcxx::rank_me()==0) {
+        std::cout<<"========= GPU CONFIGURATION OPTIONS ========="<<std::endl;
+        std::cout<<"-gpu_mem: "<<(symPACK::gpu_alloc_size/(double)(1<<20))<<" MiB"<<std::endl;
+        std::cout<<"-gpu_blk: "<<optionsFact.gpu_block_limit<<" bytes"<<std::endl;
+        std::cout<<"-trsm_limit: "<<optionsFact.trsm_limit<<" nonzeros"<<std::endl;
+        std::cout<<"-potrf_limit: "<<optionsFact.potrf_limit<<" nonzeros"<<std::endl;
+        std::cout<<"-gemm_limit: "<<optionsFact.gemm_limit<<" nonzeros"<<std::endl;
+        std::cout<<"-syrk_limit: "<<optionsFact.syrk_limit<<" nonzeros"<<std::endl;
+    }
+    symPACK::cpu_ops.insert({"trsm", 0});
+    symPACK::cpu_ops.insert({"potrf", 0});
+    symPACK::cpu_ops.insert({"syrk", 0});
+    symPACK::cpu_ops.insert({"gemm", 0});
+    symPACK::gpu_ops.insert({"trsm", 0});
+    symPACK::gpu_ops.insert({"potrf", 0});
+    symPACK::gpu_ops.insert({"syrk", 0});
+    symPACK::gpu_ops.insert({"gemm", 0});
+  }
+
+}
+#endif
 
 
 extern "C"
 int symPACK_Init(int *argc, char ***argv){
   int retval = 1;
-  // init UPC++
+  // init MPI, if necessary
+  MPI_Initialized(&symPACK::mpi_already_init);
+  if (!symPACK::mpi_already_init) MPI_Init(argc, argv);
+  int iam = 0;
+  MPI_Comm_rank(MPI_COMM_WORLD, &iam);
+  if ( symPACK::world_comm == MPI_COMM_NULL ) MPI_Comm_split(MPI_COMM_WORLD, 0, iam, &symPACK::world_comm);
+  MPI_Barrier(MPI_COMM_WORLD);
+   // init UPC++
   if ( ! libUPCXXInit ) {
     upcxx::init();
     upcxx::liberate_master_persona();
     symPACK::capture_master_scope();
     libUPCXXInit = true;
   }
-  
-  // init MPI, if necessary
-  MPI_Initialized(&symPACK::mpi_already_init);
-  if (!symPACK::mpi_already_init) MPI_Init(argc, argv);
-  if ( symPACK::world_comm == MPI_COMM_NULL ) MPI_Comm_split(MPI_COMM_WORLD, 0, upcxx::rank_me(), &symPACK::world_comm);
```

Comment:
@Hooninator I just noticed this change, converting symPACK to init MPI before UPC++.

We support either order on platforms like the HPE Cray EX, although there can be some subtle interactions on other platforms so all else being equal [we usually recommend UPC++ be init first](https://bitbucket.org/berkeleylab/upcxx/wiki/docs/mpi-hybrid.md).

Do you recall why this change was made? 

Regardless of init order, the final line above that renumbers MPI_COMM_WORLD to ensure symPACK::world_comm matches the rank ordering of UPC++ should remain where it was (after both libraries are init) and unchanged (otherwise it defeats the purpose).

Reply:
After further investigation, I believe this was probably changed to workaround [bug 4638](https://upc-bugs.lbl.gov/bugzilla/show_bug.cgi?id=4638), which is believed to arise from a defect in Cray MPI that is deployed on HPE Cray EX Slingshot-11 systems.

There are other (envvar-based) workarounds available for that bug. Applying the symPACK-level source change has the advantage of more user-friendly behavior on Slingshot systems (no user-driven workaround needed), but may potentially complicate/prevent spawning on less integrated systems (e.g. using GASNet ssh-spawner). 

On balance I think I'm inclined to leave the init order proposed in this PR (MPI then GASNet) at least until bug 4638 has a better resolution, although we might revisit that decision in the future if it creates too many problems.

In any case I've just pushed 2838378 to repair the MPI rank renumbering regression mentioned above.

CC: @PHHargrove